alf/algorithms/merlin_algorithm.py

```diff
@@ -0,0 +1,441 @@
+
+    def _variables(self):
+        return _collect_variables(self._encoders, self._decoders, self._rnn,
+                                  self._key_net, self._vae)
+
+    @property
+    def memory(self):
+        """Return the external memory of this module."""
+        return self._memory
+
+    def encode_step(self, inputs, state: MBPState):
+        """Calculate latent vector.
+
+        Args:
+            inputs (tuple): a tuple of (observation, prev_action)
+            state (MBPState)
+        """
+        observation, prev_action = inputs
+        self._memory.from_states(state.memory)
+
+        prev_action = self._action_encoder(prev_action)
+
+        prev_rnn_input = tf.concat(
+            [state.latent_vector, prev_action, state.mem_readout], axis=-1)
+
+        prev_rnn_output, prev_rnn_state = self._rnn(prev_rnn_input,
+                                                    state.rnn_state)
+
+        prev_mem_readout = self._memory.genkey_and_read(
+            self._key_net, prev_rnn_output)
+
+        self._memory.write(tf.stop_gradient(state.latent_vector))
+
+        prior_input = tf.concat([prev_rnn_output, prev_mem_readout], axis=-1)
+
+        current_input = tf.nest.map_structure(
+            lambda encoder, obs: encoder(obs)[0], self._encoders, observation)
+        current_input = tf.concat(tf.nest.flatten(current_input), axis=-1)
+
+        latent_vector, kld = self._vae.sampling_forward((prior_input,
+                                                         current_input))
+
+        next_state = MBPState(
+            latent_vector=latent_vector,
+            mem_readout=prev_mem_readout,
+            rnn_state=prev_rnn_state,
+            memory=self._memory.states)
+
+        return latent_vector, kld, next_state
+
+    def decode_step(self, latent_vector, observations):
+        """Calculate decoding loss."""
+        decoders = tf.nest.flatten(self._decoders)
+        observations = tf.nest.flatten(observations)
+        decoder_losses = [
+            decoder.train_step((latent_vector, obs)).info
+            for decoder, obs in zip(decoders, observations)
+        ]
+        loss = tf.add_n([decoder_loss.loss for decoder_loss in decoder_losses])
+        decoder_losses = tf.nest.pack_sequence_as(self._decoders,
+                                                  decoder_losses)
+        return LossInfo(loss=loss, extra=decoder_losses)
+
+    def train_step(self, inputs, state: MBPState):
+        """Train one step.
+
+        Args:
+            inputs (tuple): a tuple of (observation, action)
+        """
+        observation, _ = inputs
+        latent_vector, kld, next_state = self.encode_step(inputs, state)
+
+        # TODO: decoder for action
+        decoder_loss = self.decode_step(latent_vector, observation)
+
+        return AlgorithmStep(
+            outputs=latent_vector,
+            state=next_state,
+            info=LossInfo(
+                loss=decoder_loss.loss + kld,
```

Comment:
same here, should support weighted sum

Reply:
This weight will be handled by loss_weight in Algorithm